Add Predis support

mtk3d opened the pull request:
Hi
I've introduced changes to support Predis as an alternative Redis connection adapter, while maintaining full backward compatibility and avoiding code duplication.
This was achieved by introducing a minimal RedisClient interface, designed according to the YAGNI principle. It only exposes the necessary methods required by the prometheus adapter.
AbstractRedis.php is basically all redis generic logic moved from Redis.php

I know there already is a [PR](https://github.com/PromPHP/prometheus_client_php/pull/127) with predis but it seems abandoned.
I see [there is still need](https://github.com/PromPHP/prometheus_client_php/issues/156) to have support for predis, also I need one for my project. So I decided to make this one, and I’m ready to help bring it across the finish line 😄 

All tests are passing locally, and the code style checks are green, so I hope there will be no issues in pipeline.

mtk3d committed:
feature: add predis to dev deps

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d committed:
feature: implement redis client interface

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d committed:
feature: implement predis and phpredis clients

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d committed:
feature: add predis test

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d committed:
refactor: yagni for interface

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d committed:
refactor: use abstract redis and keep full backward compatibility

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d committed:
fix: add phpdocs

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d committed:
fix: bring _prefix() back

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d committed:
fix: fix prefix for PHPRedis

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d committed:
fix: run phpcs

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d committed:
tests: add config for blackbox test

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d committed:
fix: remove all redis notes from predis

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d committed:
fix: estabish connection after first call in predis

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d committed:
fix: add exception handling for predis connection

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d commented:
Ok, it's ready now

mtk3d committed:
chore: remove unecessary comment

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d committed:
fix: remove redundant replicate commands

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d committed:
fix: remove unnecessary comparison

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d committed:
fix: remove dead code

This code has been mentioned as issue because of recurency call. Afeter
quick analyse, I've realised that this is dead code, and it's not used
anywhere.

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d commented on src/Prometheus/Storage/RedisClients/Predis.php:
```diff
@@ -0,0 +1,143 @@
+
+    public function hSetNx(string $key, string $field, mixed $value): bool
+    {
+        return $this->hSetNx($key, $field, $value);
```
I was curious why no tests catched it. After checking references of this method, I realised it's a dead code, so I'm removing it

mtk3d commented on src/Prometheus/Storage/RedisClients/Predis.php:
```diff
@@ -0,0 +1,143 @@
+
+    public function setNx(string $key, mixed $value): void
+    {
+        $this->client->setnx($key, $value) === 1;
```
Good catch, I'm removing it

mtk3d committed:
rafactor: remove OPT_READ_TIMEOUT as it's phpredis internals

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d committed:
refactor: make opt truely abstract

https://github.com/PromPHP/prometheus_client_php/pull/186/changes#r2953935115
Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d commented on src/Prometheus/Storage/RedisClients/RedisClient.php:
```diff
@@ -0,0 +1,49 @@
+
+namespace Prometheus\Storage\RedisClients;
+
+interface RedisClient
```
That's really good catch. Thank you! Will fix it

mtk3d commented on src/Prometheus/Storage/AbstractRedis.php:
```diff
@@ -0,0 +1,626 @@
+
+        $this->redis->eval(
+            <<<'LUA'
+redis.replicate_commands()
```
I've just removed `redis.replicate_commands()` for now

mtk3d committed:
fix: apply stripPrefix only to phpredis implementation

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d committed:
fix: normalise the return type of get for predis and phpredis

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d commented on src/Prometheus/Storage/AbstractRedis.php:
```diff
@@ -0,0 +1,626 @@
+        foreach ($keys as $metaKeyWithPrefix) {
+            $metaKey = $this->removePrefixFromKey($metaKeyWithPrefix);
+            $rawSummary = $this->redis->get($metaKey);
+            if ($rawSummary === false) {
```
Normalizing sounds pretty good. I'll go this way

mtk3d commented on src/Prometheus/Storage/AbstractRedis.php:
```diff
@@ -0,0 +1,626 @@
+        sort($keys);
+        $histograms = [];
+        foreach ($keys as $key) {
+            $raw = $this->redis->hGetAll(ltrim($key, $this->redis->getOption(RedisClient::OPT_PREFIX) ?? ''));
```
`stripPrefix()` method will be good approach I think

mtk3d committed:
fix: always check if connection exists

This is also required by PHPStan

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d committed:
fix: fix mixed type of getOption

For some reason PHPStan cannot recognize that this method is returning
mixed, and still trying to use int in this code, so this function would
always fall in to if statement.

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d committed:
cicd: add predis to blackbox tests

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d committed:
fix: fix test name case

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d commented:
Thanks for checking and all suggestions. I've applied them and also rebased main branch

simPod commented:
Ill check it tmrw. Make sure CI passes please. You can open secondary PR against your fork to be able to trigger actions yourself if needed.

mtk3d committed:
fix: fix failing tests

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

simPod commented on src/Prometheus/Storage/Predis.php:
```diff
@@ -0,0 +1,95 @@
+     * @param  mixed[]  $options
+     */
+    public function __construct(array $parameters = [], array $options = [])
+    {
```
This accesses Predis internal option properties directly (`$clientOptions->aggregate`, etc.). If Predis changes its options API in a major version, this breaks.

Also, `new PredisClient(self::$defaultParameters, $options, $client)` passes both default parameters AND the existing client — but the parameters are unused when a client is already injected. This is confusing; consider making the `PredisClient` constructor clearer about which args matter when an existing client is provided.

mtk3d replied:
I've refactorized this adapter, and now it's touching only as much internals as required

simPod commented on src/Prometheus/Storage/RedisClients/Predis.php:
```diff
@@ -0,0 +1,147 @@
+     */
+    public static function create(array $parameters, array $options): self
+    {
+        return new self($parameters, $options);
```
`getOption()` reads from the `$this->options` array passed at construction, while the PHPRedis adapter reads from `$this->redis->getOption()` (live client state). If the prefix is modified on the Predis client after construction, this won't reflect it. This matters for `wipeStorage()` and `removePrefixFromKey()` which rely on `getOption(OPT_PREFIX)`.

mtk3d replied:
I've replaced getOption() with getPrefix(), as it's the only option is used anywhere

simPod commented on tests/Test/Prometheus/Predis/SummaryTest.php:
```diff
@@ -0,0 +1,25 @@
+    public function configureAdapter(): void
+    {
+        $this->adapter = new Predis(['host' => REDIS_HOST]);
+        $this->adapter->wipeStorage();
```
This is a no-op override — just calls `parent::itShouldObserveWithLabels()`. Remove it.

simPod reviewed (changes requested):
Overall LGTM, few more stuff to check

mtk3d committed:
fix: add return type of keys

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d committed:
style: add comment to document intentionally sharing prefix

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d committed:
fix: bring back replicate_commands() to keep support for older redis version

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d committed:
fix: pass previous exception

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d committed:
fix: connect predis earlier to fail fast same as phpredis

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d committed:
refactor: create Predis\Client eagerly in constructor

That allowed me to simplify this adapter as well as removing some
confusing parts of code, where some settings are required but not used

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d commented on src/Prometheus/Storage/AbstractRedis.php:
```diff
@@ -0,0 +1,625 @@
+
+        $this->redis->eval(
+            <<<'LUA'
+local cursor = "0"
```
That's right, I'm restoring it 👍

mtk3d commented on src/Prometheus/Storage/RedisClients/Predis.php:
```diff
@@ -0,0 +1,147 @@
+
+    public function del(array|string $key, string ...$other_keys): void
+    {
+        $this->getClient()->del($key, ...$other_keys);
```
I've added connect() to connection ensure method, and I've added a test to verify if behavior is the same for both adapters

mtk3d commented on src/Prometheus/Storage/RedisClients/Predis.php:
```diff
@@ -0,0 +1,147 @@
+        $this->getClient()->del($key, ...$other_keys);
+    }
+
+    /**
```
Added 👍

mtk3d commented on src/Prometheus/Storage/RedisClients/RedisClient.php:
```diff
@@ -0,0 +1,45 @@
+    /**
+     * @return string[]
+     */
+    public function keys(string $pattern);
```
Added 👍

mtk3d committed:
refactor: remove no-op override

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d committed:
refactor: replase universal getOption() to simple getPrefix()

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d committed:
fix: add type to inform phpstan

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d committed:
readme: add information about new predis adapter

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d commented:
It's ready now I think

simPod commented on src/Prometheus/Storage/RedisClients/PHPRedis.php:
```diff
@@ -0,0 +1,173 @@
+
+        if (isset($this->options['database'])) {
+            $this->redis->select($this->options['database']);
+        }
```
Missing `$e` as previous exception — this is a regression from the original `Redis.php` which passed it as third arg.

```suggestion
                $e->getCode(),
                $e
```

simPod commented on src/Prometheus/Storage/RedisClients/PHPRedis.php:
```diff
@@ -0,0 +1,173 @@
+    }
+
+    public function setNx(string $key, mixed $value): void
+    {
```
Original `\RedisException` is lost. Should preserve the cause chain:

```suggestion
            throw new RedisClientException($e->getMessage(), $e->getCode(), $e);
```

simPod commented on README.md:
````diff
@@ -70,7 +74,8 @@ header('Content-type: ' . RenderTextFormat::MIME_TYPE);
 ```
 
-Change the Redis options (the example shows the defaults):
+change the Redis options (the example shows the defaults):
````
Nit: accidentally lowercased the heading.

```suggestion
Change the Redis options (the example shows the defaults):
```

mtk3d replied:
Fixed, thanks

simPod reviewed (changes requested):
Close to mergeable. One regression from the code move and a couple minor items.

simPod commented:
Also I believe @LKaemmerling needs DCO check to pass.

simPod reviewed (dismissed):
LGTM like this

mtk3d committed:
chore: fix regression in passing previous exception if thrown

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d committed:
docs: capital letter typo

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d committed:
docs: update shared prefix note

Signed-off-by: Mateusz Cholewka <mateusz@cholewka.com.pl>

mtk3d commented:
Ok, done. Thank you @simPod for detailed review! I really appreciate your work.
All comments have been addressed (maybe except one I'm not sure about).
I've also signed all commits.

LKaemmerling reviewed (approved):
Looks really good, thank you a lot! Great work @mtk3d (and thanks @simPod as well!)
